# Praktikum 4 
## Klasifikasi dengan AAN

### Data Pra-processing
**Langkah 1 - Import Libary**

In [25]:
# Library untuk melakukan komputasi numerik
import numpy as np
# library untuk manipulasi dan analisis data
import pandas as pd
# Library untuk pengembangan model machine learning dan deep learning
import tensorflow as tf

**Langkah 2 - Load Data**

In [26]:
# Membaca dataset dari file CSV
dataset = pd.read_csv('./Dataset/Churn_Modelling.csv')
# Mengambil fitur (X) dari dataset. Di sini, "3:-1" mengambil kolom mulai dari indeks ke-3 hingga kolom sebelum indeks terakhir
X = dataset.iloc[:, 3:-1].values
# Mengambil label (y) dari dataset. Di sini, "-1" mengambil kolom terakhir sebagai label
y = dataset.iloc[:, -1].values

In [27]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


### Penjelasan :
- Setiap baris dalam dataset mewakili entitas tertentu atau catatan (misalnya, pelanggan dalam sebuah bank).
- Setiap kolom dalam dataset mewakili atribut atau fitur dari entitas tersebut. 
- Nilai-nilai dalam dataset ini adalah data aktual yang diambil dari entitas tersebut. Contohnya, baris pertama menunjukkan bahwa pelanggan memiliki nomor pelanggan 619, berasal dari Prancis, jenis kelaminnya adalah perempuan, memiliki skor kredit tertentu, saldo akun tertentu, dan sebagainya.

**Langkah 3 - Encoding Data Kategorikal**

In [28]:
from sklearn.preprocessing import LabelEncoder # Import library LabelEncoder dari scikit-learn
le = LabelEncoder() # Inisialisasi objek LabelEncoder
X[:, 2] = le.fit_transform(X[:, 2]) # Melakukan encoding pada kolom ke-2 (indeks 2) dari matriks fitur X

In [29]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


### Penjelasan :
Setelah encoding kategorikal, hasilnya adalah bahwa data kategorikal dalam kolom yang dienkoding digantikan oleh bilangan bulat. Contohnya, jika kolom jenis kelamin semula memiliki nilai 'Female' dan 'Male', setelah encoding, 'Female' mungkin diganti menjadi 0 dan 'Male' menjadi 1. 

**Langkah 4 - Encoding Kolom "Geography" dengan One Hot Encoder**

In [30]:
# Import library ColumnTransformer dan OneHotEncoder dari scikit-learn
from sklearn.compose import ColumnTransformer # untuk menerapkan berbagai transformasi pada kolom-kolom tertentu dalam matriks fitur.
from sklearn.preprocessing import OneHotEncoder # untuk melakukan one-hot encoding pada data kategorikal

# Inisialisasi objek ColumnTransformer
# Mengubah kolom dengan indeks 1 (kolom "geography") menggunakan OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')

# Melakukan transformasi pada matriks fitur X
X = np.array(ct.fit_transform(X))

In [31]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Penjelasan:
Hasilnya setelah encoding kolom "geography" dengan OneHotEncoder adalah kolom "geography" digantikan oleh beberapa kolom tambahan yang mewakili setiap kategori unik dalam kolom tersebut. Setiap kolom tambahan ini mengandung nilai biner (0 atau 1) yang menunjukkan apakah suatu baris data termasuk dalam kategori tersebut atau tidak.

**Langkah 5 - Split Data**

In [32]:
from sklearn.model_selection import train_test_split # Library untuk membagi dataset menjadi data latih dan data uji dengan proporsi tertentu.

# Memisahkan dataset menjadi data latih (train) dan data uji (test)
# data dipecah menjadi X_train, X_test, y_train, dan y_test
# dengan proporsi data uji sebesar 20% (test_size=0.2) dari keseluruhan dataset
# serta mengatur seed (random_state) agar hasil pengujian dapat direproduksi
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

**Langkah 6 - Scaling Fitur**

In [33]:
from sklearn.preprocessing import StandardScaler #Library untuk melakukan penskalaan data sehingga memiliki mean 0 dan deviasi standar 1.
sc = StandardScaler() # Inisialisasi objek StandardScaler
X_train = sc.fit_transform(X_train) # Melakukan penskalaan (scaling) data latih (X_train) menggunakan StandardScaler
X_test = sc.transform(X_test) # Melakukan penskalaan (scaling) data uji (X_test) menggunakan StandardScaler

## Membuat Model ANN

**Langkah 1 - Inisiasi Model ANN**

In [34]:
ann = tf.keras.models.Sequential() # Membuat objek Sequential untuk model ANN

**Langkah 2 - Membuat Input Layer dan Hidden Layer Pertama**

In [35]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) # Menambahkan layer Dense pertama dengan 6 unit neuron dan fungsi aktivasi ReLU

**Langkah 3 - Membuat Hidden Layer Kedua**

In [36]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) # Menambahkan layer Dense kedua dengan 6 unit neuron dan fungsi aktivasi ReLU

**Langkah 4 - Membuat Output Layer**

In [37]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) # Menambahkan layer Dense ketiga dengan 1 unit neuron dan fungsi aktivasi sigmoid

## Training Model

**Langkah 1 - Compile Model (Menyatukan Arsitektur) ANN**

In [38]:

# Compile model ANN dengan konfigurasi
# - 'adam' sebagai optimizer,
# - 'binary_crossentropy' sebagai fungsi loss (untuk masalah klasifikasi biner),
# - 'accuracy' sebagai metrik yang akan dipantau selama pelatihan.
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

**Langkah 2 - Fitting Model**

In [39]:

# Melakukan pelatihan model ANN menggunakan data latih
# - batch_size=32: Ini adalah jumlah sampel yang digunakan pada setiap iterasi pembelajaran.
# - epochs=100: Ini adalah jumlah iterasi (epoch) selama pelatihan.
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
161/250 [==================>...........] - ETA: 0s - loss: 0.6427 - accuracy: 0.6990

250/250 [==============================] - 1s 2ms/step - loss: 0.6031 - accuracy: 0.7271
Epoch 2/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4878 - accuracy: 0.7854
Epoch 3/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4491 - accuracy: 0.8005
Epoch 4/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4334 - accuracy: 0.8109
Epoch 5/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4257 - accuracy: 0.8196
Epoch 6/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4202 - accuracy: 0.8253
Epoch 7/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4158 - accuracy: 0.8274
Epoch 8/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4120 - accuracy: 0.8299
Epoch 9/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4083 - accuracy: 0.8303
Epoch 10/100
250/250 [==============================] - 0s 2ms/step - loss: 0.4039 - accuracy: 0.8322
E

### Memodelkan Data Baru dan Buat Prediksi

In [40]:
# Memprediksi hasil menggunakan model ANN pada data yang telah diubah skala (scaled)
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

1/1 [==============================] - 0s 124ms/step
[[False]]


hasil diatas menunjukkan bahwa model memprediksi bahwa data yang Anda berikan adalah kelas negatif (0) berdasarkan ambang batas 0.5 yang digunakan dalam klasifikasi biner.

### Prediksi Dengan Data Testing

In [41]:
y_pred = ann.predict(X_test) # Melakukan prediksi pada data uji menggunakan model ANN
y_pred = (y_pred > 0.5) # Menggunakan ambang batas 0.5 untuk mengklasifikasikan hasil prediksi
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1)) # Mencetak hasil prediksi dan label sebenarnya (ground truth)

63/63 [==============================] - 0s 2ms/step
[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


Pada output diatas, setiap pasangan angka menunjukkan hasil prediksi (angka pertama) dan label sebenarnya (angka kedua) untuk satu sampel data uji.
Jika angka pertama adalah 0, itu berarti model memprediksi kelas negatif.
Jika angka pertama adalah 1, itu berarti model memprediksi kelas positif.
Angka kedua adalah label sebenarnya, di mana 0 mewakili kelas negatif dan 1 mewakili kelas positif.

### Cek Akurasi dan Confusion Matrix

In [42]:
# Import library confusion_matrix dan accuracy_score dari scikit-learn
from sklearn.metrics import confusion_matrix, accuracy_score
# Menghitung confusion matrix antara label sebenarnya (y_test) dan hasil prediksi (y_pred)
cm = confusion_matrix(y_test, y_pred)
# Mencetak confusion matrix
print(cm)
# Menghitung akurasi dari model dengan membandingkan hasil prediksi (y_pred) dengan label sebenarnya (y_test)
accuracy_score(y_test, y_pred)

[[1496   99]
 [ 188  217]]


0.8565

### Penjelasan:
 output diatas adalah hasil confussion matrix dimana :
- Confusion Matrix:
    - True Positives (TP): 217
    - True Negatives (TN): 1496
    - False Positives (FP): 99
    - False Negatives (FN): 188
- Akurasi (Accuracy): 0.8565

Dapat disimpulkan bahwa model ini memiliki akurasi yang cukup baik, tetapi masih ada beberapa kasus di mana model memberikan prediksi yang salah. Analisis lebih lanjut mungkin diperlukan untuk memahami faktor-faktor yang berkontribusi terhadap prediksi yang salah dan untuk meningkatkan kinerja model jika diperlukan.




